Blorente/refactor tool

blorente opened the pull request:
- Refactors some of the previous code to remove state and properly separate the model from the CSV output.
- Integrates unit test framework with the rest of the project.

fikovnik commented on src/test/scala/extraction/DeclaredImplicitsTest.scala:
```diff
@@ -0,0 +1,112 @@
+import framework.SemanticdbTest
+
+class DeclaredImplicitsTest extends SemanticdbTest {
+  checkExtraction(
```
Here is an idea - how about we use the case classes as just containers for data. This way, we could update the tests to something like

```scala
  implicit class NormalizedResult(that: Result) {
    // removes location property
    def normalizedImplicits: Set[DeclaredImplicit] = that.implicits.map(_.copy(loc = null))
  }

  checkExtraction(
    "Basic information about declared implicits",
    """
      |package dI
      |object basicInfo {
      | implicit def m(a: String): String = "3"
      |}
    """.trim.stripMargin, { res =>

      res.normalizedImplicits should contain only (
        DeclaredImplicit(
          fqn = "_root_.dI.basicInfo.m(Ljava/lang/String;)Ljava/lang/String;.",
          clazz = "_root_.scala.Predef.String#",
          typee = "(a: String): String",
          kind = "def",
          nargs = "1
        )
      )

      res.funs shouldBe empty
      res.links shouldBe empty
      res.params shouldBe empty
  })
```

fikovnik commented on src/test/scala/extraction/DeclaredImplicitsTest.scala:
```diff
@@ -0,0 +1,112 @@
+      |object basicInfo {
+      | implicit def m(a: String): String = "3"
+      |}
+    """.trim.stripMargin, { res =>
```
It becomes even nicer here:

```scala
  checkExtraction(
    "Two declared implicits with identical code have different ids",
    """
      |package dI
      |object m1 {
      | implicit val msgDeclaration: String = "World"
      |}
      |object m2 {
      | implicit val msgDeclaration: String = "World"
      |}
    """.trim.stripMargin, { res =>
      res.normalizedImplicits should contain only (
        DeclaredImplicit(
          fqn = "_root_.dI.m1.msgDeclaration.",
          clazz = "...",
          typee = "...",
          kind = "val",
          nargs = "0
        ),
        DeclaredImplicit(
          fqn = "...",
          clazz = "...",
          typee = "...",
          kind = "val",
          nargs = "0
        )
      )

      res.funs shouldBe empty
      res.links shouldBe empty
      res.params shouldBe empty
  })
```

The point is that it is dead-simple so see what is going on

fikovnik commented on src/main/scala/extractor/ExtractImplicits.scala:
```diff
@@ -0,0 +1,72 @@
+     params: Set[ImplicitParam],
+     funs: Seq[Apply],
+     links: Set[FunApplyWithImplicitParam],
+     implicits: Set[DeclaredImplicit])
```
`)` at the new line

fikovnik commented on src/main/scala/extractor/ExtractImplicits.scala:
```diff
@@ -0,0 +1,72 @@
+      case _ => ""
+    }
+
+    lazy val syntheticImplicits =
```
the for should go at the end of ` =` to make it more compact and consistent with match

fikovnik commented on src/main/scala/extractor/ExtractImplicits.scala:
```diff
@@ -0,0 +1,72 @@
+
+    lazy val syntheticApplies = ctx.index.synthetics.filter(_.names.exists(_.toString() == "apply"))
+
+    lazy val paramsFuns =
```
same here

fikovnik commented on src/main/scala/extractor/ExtractImplicits.scala:
```diff
@@ -0,0 +1,72 @@
+    val funs = paramsFuns.groupBy(_.fun).keys.toSeq
+    val links = paramsFuns.map(x => Serializables.createLink(x.param, x.fun)).toSet
+
+    lazy val declaredImplicits =
```
same here

fikovnik replied:
Is there a reason why this is a lazy val?
In general is there a reason for any of the lazy vals to be lazy?

fikovnik commented on src/main/scala/extractor/ExtractImplicits.scala:
```diff
@@ -0,0 +1,72 @@
+      for {
+        app <- ctx.tree collect {
+          case x: Term.Apply => Serializables.AppTerm(x, x.args.size, x.fun.pos.end)
+          case x: Term.Name => Serializables.AppTerm(x, 0, x.pos.end)
```
This is a detail, but could you add some brief comments explaining the two cases? Perhaps with a snippet what they match? This will be super helpful for others reading the code

fikovnik commented on src/main/scala/extractor/ExtractImplicits.scala:
```diff
@@ -0,0 +1,72 @@
+        syn -> Serializables.createImplicitParam(ctx, symbol, den)
+      }
+
+    lazy val syntheticApplies = ctx.index.synthetics.filter(_.names.exists(_.toString() == "apply"))
```
A comment would be nice here - with a code snippet

fikovnik commented on src/main/scala/extractor/ExtractImplicits.scala:
```diff
@@ -0,0 +1,72 @@
+
+import scala.meta._
+
+object ExtractImplicits {
```
To me the whole object is essentially a function. This can be made explicit:
```scala
object ExtractImplicit extends (SemanticCtx => Result) { ... }
```

fikovnik commented on src/main/scala/extractor/Main.scala:
```diff
@@ -12,7 +12,7 @@ object Main {
         println(s"Root: ${conf.root}, Multiple projects: ${conf.multidir}")
         val walker = new SingleProjectWalker(conf.root)
-        val results = ImplicitParamsToCSV(walker)
+        val results = walker.run(ExtractImplicits.apply)
```
Once you made the ExtractImplicit a function, you do no longer need the apply here

fikovnik commented on src/main/scala/extractor/Main.scala:
```diff
@@ -12,7 +12,7 @@ object Main {
       case Some(conf) => {
         println(s"Root: ${conf.root}, Multiple projects: ${conf.multidir}")
```
sounds like a debugging stuff - for that you can use the Logging trait (cf. SemanticsdbTest)

fikovnik replied:
Another style is to write:

```scala
object Main extends App {
  private def run(conf: Config) {
         val walker = new SingleProjectWalker(conf.root)
         val results = ImplicitParamsToCSV(walker)
         val results = walker.run(ExtractImplicits)
         CSV.dumpFiles(conf.root, results)
  }

 Cli(args) map run
}
```

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -4,15 +4,16 @@ import java.nio.file.{Files, Path}
 import java.util
 import java.util.concurrent.{ConcurrentLinkedQueue, CopyOnWriteArrayList}
 
+import extractor.ExtractImplicits.Result
```
A good time to move to a real namespace: `cz.cvut.fit.prl.scalaimplicit`

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -132,24 +133,26 @@ case class SemanticCtx(database: Database, projectPath: AbsolutePath) {
 }
 
 object SemanticDBFileVisitor {
```
This is also a function `((Path, SemanticCtx) => Result)`

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -132,24 +133,26 @@ case class SemanticCtx(database: Database, projectPath: AbsolutePath) {
 
 object SemanticDBFileVisitor {
-  def apply[T](filePath: Path, projectPath: AbsolutePath, results: ConcurrentLinkedQueue[(String, T)], f: SemanticCtx => Iterable[(String, T)]) = {
+  def apply[T](filePath: Path, f: SemanticCtx => Result): Result = {
```
What is the `T` good for?

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -132,24 +133,26 @@ case class SemanticCtx(database: Database, projectPath: AbsolutePath) {
-        f(ctx).foreach(results.add)
+        val ctx = SemanticCtx(mdb)
+        val res = f(ctx)
         print(".")
```
I think it is better to add the `Logging` trait and add a real `logger.debug(s"Processing $filePath")` message

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -132,24 +133,26 @@ case class SemanticCtx(database: Database, projectPath: AbsolutePath) {
         case NonFatal(e) =>
           val st = e.getStackTrace
           e.setStackTrace(st.take(10))
           e.printStackTrace()
```
I would just print the whole stack trace and use the logger for it

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -132,24 +133,26 @@ case class SemanticCtx(database: Database, projectPath: AbsolutePath) {
 
-abstract class SemanticDBWalker {
-  def run[T](f: SemanticCtx => Iterable[(String, T)]): ConcurrentLinkedQueue[(String, T)]
+abstract class TreeWalker {
```
This seem like a function as well.
Moreover does not have any state so it can be a trait.

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -162,27 +165,51 @@ abstract class SemanticDBWalker {
+    def mergeById[A <: ResultElement](one: Set[A], other: Set[A]): Set[A] = {
+      (one ++ other)
+        .groupBy(_.id)
+        .map{case (_, v) => v.head}
```
spaces around {}

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -162,27 +165,51 @@ abstract class SemanticDBWalker {
+    val results = Files
         .walk(root.toNIO)
         .iterator()
         .asScala
```
Perhaps you can import just this one

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -162,27 +165,51 @@ abstract class SemanticDBWalker {
         .walk(root.toNIO)
         .iterator()
         .asScala
         .filter { file =>
```
This will fit one line

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -162,27 +165,51 @@ abstract class SemanticDBWalker {
-      files.foreach(x => SemanticDBFileVisitor(x, root, results, f))
-      println(" ")
-      results
+        .map {file => SemanticDBFileVisitor(file, f)}
```
spaces

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -162,27 +165,51 @@ abstract class SemanticDBWalker {
-      println(" ")
-      results
+        .map {file => SemanticDBFileVisitor(file, f)}
+        .fold(Result(Set(), Seq(), Set(), Set())) {(acc, partial) =>
```
I have seen this before: `Result(Set(), Seq(), Set(), Set())`

How about to create a constant in `Result` like:
```
object Result {
  val Empty = Result(Set(), Seq(), Set(), Set())
}
```

so you can write:

`fold(Result.Empty)(mergeResult)`

fikovnik commented on src/main/scala/extractor/SemanticCtx.scala:
```diff
@@ -162,27 +165,51 @@ abstract class SemanticDBWalker {
+        .fold(Result(Set(), Seq(), Set(), Set())) {(acc, partial) =>
+          mergeResults(acc, partial)
+        }
+    println(" ")
```
? :-)

blorente replied:
¯\_(ツ)_/¯
In all seriousness, sorry about that.

fikovnik commented on src/main/scala/extractor/Serializables.scala:
```diff
@@ -2,85 +2,102 @@ package extractor
-
-  override val csvHeader: Seq[String] = Seq("sourcelink", "path", "line", "col", "code", "symbol", "fqfn", "fqparamlist", "nargs")
-  override val csvValues: Seq[String] = Seq(id, file, line, col, code, symbol, "", "", params.toString)
+final case class Location(path: String, line: String, col: String) {
```
Why is line and col a String?

blorente replied:
Because we need to clone them in order to not keep too many references to the tree, it was a performance issue once. Of course it can be cloned into Ints, but I guessed that for the time it reaches the case class we shouldn't perform any operation on the data anyways, so it didn't really matter.

fikovnik commented on src/main/scala/extractor/Serializables.scala:
```diff
@@ -2,85 +2,102 @@ package extractor
+  def createSyntheticApplication(ctx: SemanticCtx, synth: Synthetic, file: String, params: Int): Apply = {
+    Apply(
+      location = Location(file, synth.position.endLine.toString, synth.position.endColumn.toString),
+      code = s"apply(${
```
IMHO it not good to include logic in string interpolation. I would write it as
```scala
s"apply(${Seq.fill(param)("_").mkString(", ")}")`
```

fikovnik commented on src/test/scala/extraction/FQNTest.scala:
```diff
@@ -0,0 +1,118 @@
+    res.implicits.size shouldBe 2
+    res.normalizedImplicits should contain only (
+      DeclaredImplicit(
+        location = dummyLocation,
```
How about `Location.Empty`

fikovnik commented on src/test/scala/extraction/FQNTest.scala:
```diff
@@ -0,0 +1,118 @@
+      | implicit def m(a: String): String = "3"
+      |}
+    """.trim.stripMargin, { res =>
+      res.implicits.size shouldBe 1
```
I think the matcher is like `should have size(1)`

fikovnik commented on src/test/scala/extraction/FQNTest.scala:
```diff
@@ -0,0 +1,118 @@
+      def find(n: String): DeclaredImplicit = {
+        res.implicits.find(_.fqn.contains(n)).get
+      }
+      res.implicits.size shouldBe 8
```
Here can be the map of fqns and then should contain only.

fikovnik commented on src/test/scala/extraction/FunctionApplicationTest.scala:
```diff
@@ -0,0 +1,74 @@
+      | message1Arg("Hello")
+      |}
+    """.trim.stripMargin, { res =>
+    res.funs.size shouldBe 2
```
Can this be matched similarly to implcits?

fikovnik commented on src/test/scala/extraction/FunctionApplicationTest.scala:
```diff
@@ -0,0 +1,74 @@
+      | say("Hello")("World")
+      |}
+    """.trim.stripMargin, { res =>
+    res.funs.size shouldBe 1
```
same here

fikovnik commented on src/test/scala/framework/SemanticdbTest.scala:
```diff
@@ -91,7 +95,7 @@ abstract class SemanticdbTest extends FunSuite with Matchers with LazyLogging {
     Files.write(testFile.toPath, code.getBytes)
 
     try {
-      logger.debug(s"Compiling $testFile with "+g.settings.toString())
```
It can be removed, true.

fikovnik commented on src/test/scala/heuristics/ClassificationTest.scala:
```diff
@@ -0,0 +1,60 @@
+
+class ClassificationTest extends SemanticdbTest {
+
+
```
extra lines

fikovnik replied:
What would be great is to actually add the scalafmt to the build.

fikovnik commented on src/test/scala/heuristics/ClassificationTest.scala:
```diff
@@ -0,0 +1,60 @@
+    * @param wrapperFunctions
+    * @return
+    */
+  def classifyUsages(ctx: SemanticCtx, wrapperFunctions: Seq[Synthetic]): Wrappers = {
```
Not sure I understand this? Why is this in a test?

blorente replied:
Those (everything in `heuristics`) are exploratory tests, not really testing any functionality in the extractor, but exploring and demonstrating different classification techniques we can reimplement later. They are in the tests, because they provide some way to test our classification logic in executable code.

That said, if you are confused it means that the comment I wrote wasn't clear enough, I'll re-document it.

fikovnik commented on src/test/scala/heuristics/ImplicitClassExtensionTest.scala:
```diff
@@ -0,0 +1,134 @@
+    */
+  def checkDefinition(ctx: SemanticCtx): Unit = {
+    val res = ExtractImplicits(ctx)
+    res.implicits should have size 1
```
yay! :-)

fikovnik commented on src/test/scala/heuristics/ImplicitClassExtensionTest.scala:
```diff
@@ -0,0 +1,134 @@
+    // See that the inserted call matches the expected fqn.of.constructor(*) pattern
+    val usages = ctx.index.synthetics.filter(_.text.matches("""(\.?[\[\w\]]*)+\(\*\)"""))
+    usages should not be empty
+    usages.foreach { usage =>
```
Isn't the same thing above?

fikovnik replied:
I;m not really getting this, but cannot it be simplified a bit with just plain matching of case classes?

blorente replied:
It probably can, yes, but I was trying to imagine what a possible R script would look like that classified all the data.

In this particular instance, it is trying to identify the usage of an implicit class conversion, as described in the last paragraph of this section of the wiki: https://github.com/PRL-PRG/scalafix-rule-workshop/wiki/Patterns:-Heuristics-and-intuitions#application

blorente commented:
Fixed most of the mistakes from the last review. Unfortunately, adding scalafmt touched many files, and thus the comments are marked as outdated, so I'll try to address them here.
## 1
Regarding the formatting, I have added a pre-commit [git hook](http://githooks.com/) to the repo, that runs the scalafmt checker on all the relevant parts of the project. Of course the installation of the hook is optional, and it explained [in the readme](https://github.com/PRL-PRG/scalafix-rule-workshop/tree/blorente/refactor-tool#for-contributors).

## 2
Regarding some string matching and unusual tests in `heuristics`, those  are exploratory tests. The intention is not really to validate any functionality in the extractor, but to explore different classification techniques we can reimplement later. They are in the tests, because they provide some way to test our classification logic in executable code.

For example, `checkUsages` in `ImplicitClassExtensionTest.scala`, it is trying to identify the usage of an implicit class conversion, as described in the last paragraph of this section of the wiki: https://github.com/PRL-PRG/scalafix-rule-workshop/wiki/Patterns:-Heuristics-and-intuitions#application

## 3
Regarding `line` and `col` being `String`s in `Location`. We need to clone them in order to not keep too many references to the tree, it was a performance issue. Of course it can be cloned into Ints, but I guessed that for the time it reaches the case class we shouldn't perform any operation on the data anyways, so it didn't really matter.

fikovnik commented:
Sounds good!

Can you just elaborate the last issue - with String/Ints and performance.

fikovnik commented on src/main/scala/cz/cvut/fit/prl/scalaimplicit/extractor/ExtractImplicits.scala:
```diff
@@ -0,0 +1,124 @@
+          case x: Term.Name => Serializables.AppTerm(x, 0, x.pos.end)
+        }
+        param <- syntheticImplicits collect {
+          case (syn, den) if syn.position.end == app.term.pos.end => den
```
Can the `syntheticImplicits` be a `Map[Int, Denotation]`?

fikovnik commented on src/main/scala/cz/cvut/fit/prl/scalaimplicit/extractor/ExtractImplicits.scala:
```diff
@@ -0,0 +1,124 @@
+        param <- syntheticImplicits collect {
+          case (syn, den) if syn.position.end == app.term.pos.end => den
+        }
+        syntheticApply = syntheticApplications find { x =>
```
You should push this into the yield block:

```scala
syntheticApplications
  .find { x => x.position.end >= app.term.pos.start && x.position.end <= app.term.pos.end }
  .map { x => LinkPair(x, Serializables.createSyntheticApplication(ctx, x, file, app.params)) }
  .getOrElse(LinkPair(param, Serializables.createFunctionApplication(ctx, app, file)))
```

fikovnik commented on scripts/clean-data.py:
```diff
@@ -132,11 +166,28 @@ def clean_fqn_test():
+
 def clean_param_row(row):
     row["fqn"] = clean_fqn(row["fqn"])
     row["fqfn"] = extract_function_name(row["fqn"])
```
is the function name any different from name?

blorente replied:
yes, the function name is given in `fully.qualified.form`, and the name is just what comes after the last dot and before the first '('

fikovnik replied:
so what is `fqn`?

blorente replied:
If we eventually transition to a parameter table, we can probably do something more useful with `fqfn`

blorente replied:
`fqn` is the complete symbol `path.to.fun(param.one.type,param.two.type)return.val.type`.

fikovnik replied:
OK, here is confusion. I though that it is really just fully qualified name which together with signature makes an id.

fikovnik replied:
Wouldn't that be enough?

fikovnik commented on scripts/clean-data.py:
```diff
@@ -132,11 +166,28 @@ def clean_fqn_test():
     row["fqn"] = clean_fqn(row["fqn"])
     row["fqfn"] = extract_function_name(row["fqn"])
+    row["name"] = extract_plain_name(row["fqn"])
     row["fqparamlist"] = extract_parameter_list(row["fqn"])
```
Is this actually useful? I think the full signature is probably enough. For anything more elaborate we will need to break this down to function arguments in a new table

blorente replied:
I honestly don't know, it depends on what the usage looks like on the R side, we haven't talked about that yet.

fikovnik replied:
i think this can be easily parsed in R

fikovnik commented on scripts/mysql-schema.sql:
```diff
@@ -118,8 +117,7 @@ CREATE TABLE `declared_implicits` (
-  `type` text NOT NULL,
+  `signature` text NOT NULL,
   `kind` tinytext NOT NULL,
   `nargs` int(11) NOT NULL,
```
I think we can also remove the nargs

fikovnik replied:
Perhaps eventually, we should move to the parameter lists and parameters tables...

blorente replied:
Don't we need nargs to match the implicit conversion pattern? An implicit conversion can be an implicit def with nargs == '1'

fikovnik replied:
The problem is that nargs misses information. The following will look the same, no?
```
implicit def f(that: MyObject)
implicit def f(implicit that: MyObject)
```

blorente replied:
No, nargs ignores implicit parameters

blorente commented:
Summary of the discussion about the schema. `fqfn` and `fqparamlist` can probably be deleted, since if we need information from them we will probably have to place more sophisticated mechanisms anyway.

`nargs` holds the information about the _non-implicit_ argument number. It may be worth keeping to match declared_implicits and identify the implicit conversion pattern. Or, it may not be worth keeping, for reasons similar to those above. In any case, `nargs` appears in `funs` and `declared_implicits`. We have no use yet for the first one, and we have not implemented the second one, so it wouldn't be a big loss to remove it.

`fqn` currently holds the information about defs in the following form: `path.to.fun(path.to.param.one.type,path.to.param.two.type)path.to.return.type`. It can be discussed if further processing is necessary.

fikovnik commented:
Let's remove the `fqfn`, `fparamlist`. We can keep the `nargs` for now and I guess we can have fqn just the fully qualified name i.e. `path.to.name`

blorente commented:
So `fqn` would be everything before the first `(`, right?

fikovnik commented:
Sounds good - before `(` if there is one

blorente commented:
Should be done now
